Complier Bugs on Init

## Sourcecode

- https://github.com/WhiteMatrixTech/whiteq/blob/main/src/index.ts#L39
- https://github.com/WhiteMatrixTech/whiteq/blob/main/src/index.ts#L51

## ESLint

···
Unsafe return of type `Promise<Job<any, any, string>>` from function with return type `Promise<Job<T, R, N>>`.
···

```
{
	"resource": "/home/willin/Documents/WhiteMatrix/nft/whiteq/src/index.ts",
	"owner": "eslint",
	"code": {
		"value": "@typescript-eslint/no-unsafe-return",
		"target": {
			"$mid": 1,
			"external": "https://github.com/typescript-eslint/typescript-eslint/blob/v4.29.3/packages/eslint-plugin/docs/rules/no-unsafe-return.md",
			"path": "/typescript-eslint/typescript-eslint/blob/v4.29.3/packages/eslint-plugin/docs/rules/no-unsafe-return.md",
			"scheme": "https",
			"authority": "github.com"
		}
	},
	"severity": 8,
	"message": "Unsafe return of type `Promise<Job<any, any, string>>` from function with return type `Promise<Job<T, R, N>>`.",
	"source": "eslint",
	"startLineNumber": 39,
	"startColumn": 5,
	"endLineNumber": 39,
	"endColumn": 36
}
```

## Build

```
❯ yarn build
yarn run v1.22.10
$ tsc
src/index.ts:39:5 - error TS2322: Type 'Promise<Job<any, any, string>>' is not assignable to type 'Promise<Job<T, R, N>>'.
  Type 'Job<any, any, string>' is not assignable to type 'Job<T, R, N>'.
    Type 'string' is not assignable to type 'N'.
      'string' is assignable to the constraint of type 'N', but 'N' could be instantiated with a different subtype of constraint 'string'.

39     return q.add(name, data, opts);
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/index.ts:51:5 - error TS2322: Type 'Promise<Job<any, any, string>[]>' is not assignable to type 'Promise<Job<T, any, N>[]>'.
  Type 'Job<any, any, string>[]' is not assignable to type 'Job<T, any, N>[]'.
    Type 'Job<any, any, string>' is not assignable to type 'Job<T, any, N>'.
      Type 'string' is not assignable to type 'N'.
        'string' is assignable to the constraint of type 'N', but 'N' could be instantiated with a different subtype of constraint 'string'.

51     return q.addBulk(jobs);
       ~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
···

--- a/src/index.ts
+++ b/src/index.ts
@@ -12,9 +12,9 @@ import {
 } from './types';
 
 export class WhiteQ<T = any, R = any, N extends string = string> {
-  private queques: Map<string, Queue> = new Map();
+  private queques: Map<string, Queue<T, R, N>> = new Map();
 
-  private workers: Map<string, Worker> = new Map();
+  private workers: Map<string, Worker<T, R, N>> = new Map();
 
   private flow: FlowProducer;
 
@@ -26,15 +26,15 @@ export class WhiteQ<T = any, R = any, N extends string = string> {
     this.flow = new FlowProducer(baseOpts);
   }
 
-  public queue(queueName: string): Queue {
+  public queue(queueName: string): Queue<T, R, N> {
     if (!this.queques.has(queueName)) {
       const q = new Queue<T, R, N>(queueName, this.opts);
       this.queques.set(queueName, q);
     }
-    return this.queques.get(queueName) as Queue;
+    return this.queques.get(queueName) as Queue<T, R, N>;
   }
 
-  public addJob(queueName: string, name: string, data: T, opts: JobsOptions): Promise<Job<T, R, N>> {
+  public addJob(queueName: string, name: N, data: T, opts: JobsOptions): Promise<Job<T, R, N>> {
     const q = this.queue(queueName);
     return q.add(name, data, opts);
   }
@@ -42,7 +42,7 @@ export class WhiteQ<T = any, R = any, N extends string = string> {
   public addJobs(
     queueName: string,
     jobs: {
-      name: string;
+      name: N;
       data: T;
       opts?: BulkJobOptions;
     }[]
@@ -51,12 +51,12 @@ export class WhiteQ<T = any, R = any, N extends string = string> {
     return q.addBulk(jobs);
   }
 
-  public worker(queueName: string, processor?: string | Processor<T, R, N>, opts?: WorkerOptions): Worker {
+  public worker(queueName: string, processor?: string | Processor<T, R, N>, opts?: WorkerOptions): Worker<T, R, N> {
     if (!this.workers.has(queueName)) {
-      const w = new Worker(queueName, processor, opts);
+      const w = new Worker<T, R, N>(queueName, processor, opts);
       this.workers.set(queueName, w);
     }
-    return this.workers.get(queueName) as Worker;
+    return this.workers.get(queueName) as Worker<T, R, N>;
   }
 
   public addFlowJobs(flow: FlowJob, opts?: FlowOpts): Promise<JobNode> {

